lib/elixir/lib/task/supervised.ex

```diff
@@ -148,4 +147,194 @@ defmodule Task.Supervised do
+    try do
+      next.({:cont, []})
+    catch
+      kind, reason ->
+        stacktrace = System.stacktrace
+        stream_close(waiting, monitor_pid, monitor_ref)
+        :erlang.raise(kind, reason, stacktrace)
+    else
+      {:suspended, [value], next} ->
+        waiting = stream_spawn(value, spawned, waiting, mfa, spawn, monitor_pid, monitor_ref)
+        stream_reduce({:cont, acc}, max - 1, spawned + 1, delivered, waiting, next,
+                      reducer, mfa, spawn, monitor_pid, monitor_ref, timeout)
+      {_, [value]} ->
+        waiting = stream_spawn(value, spawned, waiting, mfa, spawn, monitor_pid, monitor_ref)
+        stream_reduce({:cont, acc}, max - 1, spawned + 1, delivered, waiting, :done,
+                      reducer, mfa, spawn, monitor_pid, monitor_ref, timeout)
+      {_, []} ->
+        stream_reduce({:cont, acc}, max, spawned, delivered, waiting, :done,
+                      reducer, mfa, spawn, monitor_pid, monitor_ref, timeout)
+    end
+  end
+
+  defp stream_deliver({:suspend, acc}, max, spawned, delivered, waiting, next,
+                      reducer, mfa, spawn, monitor_pid, monitor_ref, timeout) do
+    {:suspended, acc, &stream_deliver(&1, max, spawned, delivered, waiting, next,
+                                      reducer, mfa, spawn, monitor_pid, monitor_ref, timeout)}
+  end
+  defp stream_deliver({:halt, acc}, max, spawned, delivered, waiting, next,
+                      reducer, mfa, spawn, monitor_pid, monitor_ref, timeout) do
+    stream_reduce({:halt, acc}, max, spawned, delivered, waiting, next,
+                  reducer, mfa, spawn, monitor_pid, monitor_ref, timeout)
+  end
+  defp stream_deliver({:cont, acc}, max, spawned, delivered, waiting, next,
+                      reducer, mfa, spawn, monitor_pid, monitor_ref, timeout) do
+    case waiting do
+      %{^delivered => {kind, value}} when kind in [:ok, :exit] ->
+        try do
+          reducer.({kind, value}, acc)
+        catch
+          kind, reason ->
+            stacktrace = System.stacktrace
+            is_function(next) && next.({:halt, []})
+            stream_close(waiting, monitor_pid, monitor_ref)
+            :erlang.raise(kind, reason, stacktrace)
+        else
+          pair ->
+            stream_deliver(pair, max, spawned, delivered + 1, Map.delete(waiting, delivered), next,
+                           reducer, mfa, spawn, monitor_pid, monitor_ref, timeout)
+        end
+      %{} ->
+        stream_reduce({:cont, acc}, max, spawned, delivered, waiting, next,
+                      reducer, mfa, spawn, monitor_pid, monitor_ref, timeout)
+    end
+  end
+
+  defp stream_close(waiting, monitor_pid, monitor_ref) do
+    Process.unlink(monitor_pid)
+    Process.exit(monitor_pid, :kill)
+
+    receive do
+      {:DOWN, ^monitor_ref, _, _, _} -> :ok
+    end
+
+    for {position, {:running, pid}} <- waiting do
+      Process.unlink(pid)
+      Process.exit(pid, :kill)
```

Comment:
If we are going to unlink and kill we should do the same in `Task.shutdown`. There is a race condition where the caller can be shutdown in between `unlink/1` and `exit/2` and the tasks won't be killed if it is trapping exits. Do we care about this? IIRC a `:supervisor` follows the same pattern. We could avoid this race condition by having the `monitor_pid` kill any alive tasks when its parent exits. Would this be desirable anyway? Links should be taking care of that anyway (ofc with trapping exits it may not the case).

Do we want to kill these tasks before the `monitor_pid`? Killing is asynchronous and we should wait for the tasks to be stopped, i.e. wait for the `:DOWN`. Technically it is possible for a task to a hold a lock (registered name/ets) that will cause an error when the caller is restarted and the task it is slow to exit (because it has a lot to do/unlucky scheduling).

Given that we should also wait for `{:ok, value}` `:DOWN`s too.